Adds glob-style pattern matching for file specifications using new "include" property in tsconfig.json

rbuckton opened the pull request:
Adds support for expanding [glob](http://en.wikipedia.org/wiki/Glob_(programming))-like patterns in "include" and "exclude" properties of tsconfig.json. The following patterns are supported:

* `*` - Matches zero or more characters, excluding directory separators.
* `?` - Matches any one character, excluding directory separators.
* `[abc]` - Matches any one character in the set `{ "a", "b", "c" }`.
* `[a-z]` - Matches any one character in between `"a"` and `"z"`.
* `[!abc]` - Matches any one character **not** in the set `{ "a", "b", "c" }`.
* `**/` - Recursively matches any subdirectory.

At this time, character escape sequences are not supported, as there is no standard cross-platform convention. The standard escape sequence in a Unix shell is prefixed with `\`, which is reserved as a directory separator on Windows, and is normalized to `/` by TypeScript.

This does not use the [glob](http://npmjs.org/glob) module, as TypeScript runs in more host environments than just NodeJS, and this would need interoperate with the language service host.

```json
{
    "compilerOptions": {
        "out": "test.js"
    },
    "files": [
      "literal.ts"
    ],
    "include": [
        "**/*.ts"
    ],
    "exclude": [
        "node_modules",
        "tests/**/*.spec.ts",
        "utils/t2.ts"
    ]
}
```

* Files in the `"files"` list are always added to the list of files.
* The `"exclude"` list only applies to files matched via `"include"`.
* If neither `"files"` nor `"include"` are present, `"include"` defaults to `"**/*.ts"`. This preserves the existing behavior.

rbuckton commented:
This differs slightly from #3188, as to how to exclude a directory. The pattern matching here is expecting to match files, so to exclude a directory you would have to change the example from #3188 to the following:

```
{
    "compilerOptions": {
        "out": "test.js"
    },
    "exclude": [
        "node_modules/**/*",
        "test.ts",
        "utils/t2.ts"
    ]
}
```

It would, however, be feasible to change the implementation of `createExcludePattern` to be less restrictive.

ahejlsberg commented:
Looks interesting. The question here is whether the delta between the simple exclude list and this PR is worth another 500 lines of non-trivial code (plus we'd have to add tests for all the wildcard features). I do think the vast majority of use cases are covered with just a simple list. Also, the simple list solution has the advantage that it never even enumerates excluded subfolders whereas this code appears to enumerate them and the subsequently eliminates all the files. This could be significant from a performance perspective. And obviously it's fixable with some extra work.

rbuckton commented:
In the simple list case, if there are no wildcards in the file list, no directories are enumerated. I agree that a small change can avoid enumerating excluded directories in the case where there is a wildcard in the list, by checking the base path against the exclusion list prior to descending into a directory in both `expandWildcard` and `expandRecursiveDirectory`.

rbuckton commented:
I've made a few changes, for your consideration:

* The `"files"` property no longer supports wildcard matching, files in the `"files"` array are always added. 
* The `"exclude"` property no longer applies to `"files"`, which is consistent with #3188.
* Added an `"include"` property, which supports wildcards.
* A literal file specification in the `"include"` array will be skipped if it is not present on disk.
* If neither `"files"` nor `"include"` are present in in the file, `"include"` would have the default value of `["**/*.ts"]`, which preserves the existing behavior when `"files"` is not present.

I've also added a number of tests to cover this feature.

Zoltu commented:
Is it safe to assume that the pattern matching is relative to the directory the `tsconfig.json` file is located in?  What if the user specifies the `--rootDir` compiler option?  Will the paths be relative to the supplied root directory?

I assume the pattern matching is on the full (relative) path?  `foo.ts` would not match `bar/foo.ts`?

panuhorsmalahti commented on src/compiler/commandLineParser.ts:
```diff
@@ -389,23 +395,449 @@ module ts {
+          * @param isLiteralFile A value indicating whether the file to be added is 
+          *                      a literal file, or the result of matching a file specification.
+          */
+        function includeFile(file: string, isLiteralFile: boolean): void {
```
It's better to add the output array as a function parameter to decouple this function from expandFiles instead using a closure to access output. This increases readability, decoupling, testing and performance of this function. This same comment probably applies to other functions defined in expandFiles too.

scottwio commented:
Is this still happening seems to have stalled

mhegazy commented:
@rbuckton got busy with other changes. i would say it is open for others to pick it up if anyone is interested.

nickroberts commented:
Would love to see this, or something like the `atom-typescript` `filesGlob` property. We can't get everyone to use Atom, so this would be great if it was native.

I know the community is yearning for glob support!

Dashue commented:
+1

dsebastien commented:
+1 million, this will make things much easier for some Angular/RxJS users who don't simply have the hello world example file structure :)

chrsmrtn- commented:
+1 as well. Can hobble along without, but this would make things a lot easier for me.  Started using Aurelia which includes it's own d.ts files.  They track versions in the file path for the definition file. So I can see each time I update, i'll have to manually update all references into the framework so typescript can find the definition files.  With this in, then I can path match to ignore the version part of the path so typescript will always find the updated definition file.

glen-84 commented:
I spent over 4 hours last night trying to get this to run on top of the latest code. First with a rebase, which had about 8 or 9 conflicted files, where I was able to resolve all but one of those (commandLineParser.ts), which was just too confusing for me. Then I aborted that, and tried to reconstruct the branch manually, but I quickly ran into issues with the test system (which is what I tried to start with). So, I gave up. I just don't know enough about the code and the testing system. :(

@rbuckton, do you think that you will ever find time to finish this, or at least rebase it, or will it have to be picked up by someone else?

@chrsmrtn-, one option is to use `exclude` instead of `files`, but it's not ideal.

Codier commented:
Really need this +million

scottwio commented:
It's crazy that this is pretty much being ignored, editors are adding file globs into them and people are creating modules just to do this task clearly the community want this feature. As a work around I currently have a watch with https://github.com/wjohnsto/tsconfig-glob and tsc, you could also go down the route of grunt-ts.

rbuckton commented:
@scottwio: Actually, I'm about to close this PR and open a new one with an updated implementation.

scottwio commented:
@rbuckton in that case your my hero

rbuckton commented:
Closed in favor of #5980.
